bridges/snowbridge/primitives/router/src/inbound/v2.rs

```diff
@@ -0,0 +1,642 @@
+	},
+}
+
+/// Reason why a message conversion failed.
+#[derive(Copy, Clone, TypeInfo, PalletError, Encode, Decode, RuntimeDebug, PartialEq)]
+pub enum ConvertMessageError {
+	/// Invalid foreign ERC-20 token ID
+	InvalidAsset,
+	/// Cannot reachor a foreign ERC-20 asset location.
+	CannotReanchor,
+}
+
+pub trait ConvertMessage {
+	fn convert(
+		message: Message,
+		origin_account: Location,
+	) -> Result<(Xcm<()>, u128), ConvertMessageError>;
+}
+
+pub struct MessageToXcm<
+	EthereumNetwork,
+	InboundQueuePalletInstance,
+	ConvertAssetId,
+	GatewayProxyAddress,
+	EthereumUniversalLocation,
+	GlobalAssetHubLocation,
+> where
+	EthereumNetwork: Get<NetworkId>,
+	InboundQueuePalletInstance: Get<u8>,
+	ConvertAssetId: MaybeEquivalence<TokenId, Location>,
+	GatewayProxyAddress: Get<H160>,
+	EthereumUniversalLocation: Get<InteriorLocation>,
+	GlobalAssetHubLocation: Get<Location>,
+{
+	_phantom: PhantomData<(
+		EthereumNetwork,
+		InboundQueuePalletInstance,
+		ConvertAssetId,
+		GatewayProxyAddress,
+		EthereumUniversalLocation,
+		GlobalAssetHubLocation,
+	)>,
+}
+
+impl<
+		EthereumNetwork,
+		InboundQueuePalletInstance,
+		ConvertAssetId,
+		GatewayProxyAddress,
+		EthereumUniversalLocation,
+		GlobalAssetHubLocation,
+	> ConvertMessage
+	for MessageToXcm<
+		EthereumNetwork,
+		InboundQueuePalletInstance,
+		ConvertAssetId,
+		GatewayProxyAddress,
+		EthereumUniversalLocation,
+		GlobalAssetHubLocation,
+	>
+where
+	EthereumNetwork: Get<NetworkId>,
+	InboundQueuePalletInstance: Get<u8>,
+	ConvertAssetId: MaybeEquivalence<TokenId, Location>,
+	GatewayProxyAddress: Get<H160>,
+	EthereumUniversalLocation: Get<InteriorLocation>,
+	GlobalAssetHubLocation: Get<Location>,
+{
+	fn convert(
+		message: Message,
+		origin_account_location: Location,
+	) -> Result<(Xcm<()>, u128), ConvertMessageError> {
+		let mut message_xcm: Xcm<()> = Xcm::new();
+		if message.xcm.len() > 0 {
+			let xcm_string = hex::encode(message.xcm.clone());
+			log::info!(target: LOG_TARGET,"found xcm payload: {:x?}", xcm_string);
+			// Allow xcm decode failure so that assets can be trapped on AH instead of this
+			// message failing but funds are already locked on Ethereum.
+			if let Ok(versioned_xcm) = VersionedXcm::<()>::decode_with_depth_limit(
+				MAX_XCM_DECODE_DEPTH,
+				&mut message.xcm.as_ref(),
+			) {
+				if let Ok(decoded_xcm) = versioned_xcm.try_into() {
+					message_xcm = decoded_xcm;
+				} else {
+					log::error!(target: LOG_TARGET,"unable to decode xcm");
+				}
+			} else {
+				log::error!(target: LOG_TARGET,"unable to decode versioned xcm");
+			}
+		}
+
+		log::info!(target: LOG_TARGET,"xcm decoded as {:?}", message_xcm);
+
+		let network = EthereumNetwork::get();
+
+		// use eth as asset
+		let fee_asset = Location::new(2, [GlobalConsensus(EthereumNetwork::get())]);
+		let fee: XcmAsset = (fee_asset.clone(), message.execution_fee).into();
+		let eth: XcmAsset =
+			(fee_asset.clone(), message.execution_fee.saturating_add(message.value)).into();
```

Comment:
related to question above, seems like `message.value` is how much native eth is also included for transfer with this message, right?

Reply:
Correct.